src/audio/drc/drc_algorithm_hifi3.c

```diff
@@ -0,0 +1,469 @@
+		release_frames_f64 = AE_SRAI64(release_frames_f64, 10); /* Q36.28 -> Q46.18 */
+		AE_MULAF32R_LL(release_frames_f64, p->kB, x); /* Q20.12 * Q11.21 = Q46.18 */
+		AE_MULAF32R_LL(release_frames_f64, p->kC, x2);
+		AE_MULAF32R_LL(release_frames_f64, p->kD, x3);
+		AE_MULAF32R_LL(release_frames_f64, p->kE, x4);
+		release_frames_f64 = AE_SLAI64S(release_frames_f64, 10); /* Q46.18 -> Q36.28 */
+		release_frames = AE_ROUND32F48SSYM(release_frames_f64); /* Q36.28 -> Q20.12 */
+
+		/* db_per_frame = kSpacingDb / release_frames */
+		db_per_frame = drc_inv_fixed(release_frames, 12, 30); /* Q2.30 */
+		tmp = p->kSpacingDb << 16; /* Q16.16 */
+		db_per_frame = mult(db_per_frame, tmp, 30, 16, 24); /* Q8.24 */
+		envelope_rate = db2lin_fixed(db_per_frame); /* Q12.20 */
+	} else {
+		/* Attack mode - compression_diff_db should be positive dB */
+
+		/* Fix gremlins. */
+		if (is_bad_db)
+			compression_diff_db = ONE_Q21;
+
+		/* As long as we're still in attack mode, use a rate based off
+		 * the largest compression_diff_db we've encountered so far.
+		 */
+		tmp = AE_SLAI32S(compression_diff_db, 3); /* Q11.21 -> Q8.24 */
+		state->max_attack_compression_diff_db =
+			MAX(state->max_attack_compression_diff_db, (int32_t)tmp);
+
+		eff_atten_diff_db =
+			MAX(HALF_Q24, state->max_attack_compression_diff_db); /* Q8.24 */
+
+		/* x = 0.25f / eff_atten_diff_db;
+		 * => x = 1.0f / (eff_atten_diff_db << 2);
+		 */
+		x = drc_inv_fixed(eff_atten_diff_db, 22 /* Q8.24 << 2 */, 26); /* Q6.26 */
+		envelope_rate = AE_SUB32(ONE_Q20,
+					 drc_pow_fixed(x, p->one_over_attack_frames)); /* Q12.20 */
+	}
+
+	tmp = AE_SLAI32S(envelope_rate, 10); /* Q12.20 -> Q2.30 */
+	state->envelope_rate = tmp;
+	state->scaled_desired_gain = scaled_desired_gain;
+}
+
+/* Calculate compress_gain from the envelope and apply total_gain to compress
+ * the next output division.
+ */
+void drc_compress_output(struct drc_state *state,
+			 const struct sof_drc_params *p,
+			 int nbyte,
+			 int nch)
+{
+	const int div_start = state->pre_delay_read_index;
+	int count = DRC_DIVISION_FRAMES >> 2;
+
+	ae_f32 c, base, r, r2, r4; /* Q2.30 */
+	ae_f32 x[4]; /* Q2.30 */
+	ae_f32 post_warp_compressor_gain;
+	ae_f32 total_gain;
+	ae_f32 tmp;
+
+	int i, j, ch, inc;
+	int16_t *sample16_p; /* for s16 format case */
+	int32_t *sample32_p; /* for s24 and s32 format cases */
+	int32_t sample;
+	int is_2byte = (nbyte == 2); /* otherwise is 4-bytes */
+
+	/* Exponential approach to desired gain. */
+	if (state->envelope_rate < ONE_Q30) {
+		/* Attack - reduce gain to desired. */
+		c = AE_SUB32(state->compressor_gain, state->scaled_desired_gain);
+		base = state->scaled_desired_gain;
+		r = AE_SUB32(ONE_Q30, state->envelope_rate);
+		x[0] = mult(c, r, 30, 30, 30);
+		for (j = 1; j < 4; j++)
+			x[j] = mult(x[j - 1], r, 30, 30, 30);
+		r2 = mult(r, r, 30, 30, 30);
+		r4 = mult(r2, r2, 30, 30, 30);
+
+		i = 0;
+		inc = 0;
+		while (1) {
+			if (is_2byte) { /* 2 bytes per sample */
```

Comment:
This is a constant within the loop but it's being checked on every iteration, ideally we should have two loops types here and call the correct loop based on is_2byte (so it's only being checked once).

Reply:
Thanks, updated.